### ISSUE TITLE
Treesitter highlights get cleaned

### ISSUE THREAD

**Issue by LC0117** [EdenEast/nightfox.nvim#152]
## Description

Many treesitter highlights get cleaned when `colorscheme nightfox`

## Behavior

**Expected behavior**
Correct treesitter highlight(have run `NightfoxCompile`)

**Screenshots**
This is what happened to nightfox
<img width="702" alt="截屏2022-05-03 11 47 42" src="https://user-images.githubusercontent.com/77481149/166402474-05b4fa4b-4c6d-47af-888f-78f9e6608805.png">
When I try `hi TSFunction`, nvim gives message `TSFunction xxx cleared`, this happens also to TSKeyword.

Another theme support treesitter.
<img width="702" alt="截屏2022-05-03 11 47 55" src="https://user-images.githubusercontent.com/77481149/166402638-978ce9ed-9682-486f-82e1-636a3610cf77.png">

## Configuration

```lua
require('nightfox').setup({
    options = {
      dim_inactive = false,
      styles = {
        comments = 'italic',
        functions = 'italic',
        keywords = 'bold,italic',
      },
      inverse = {
        match_paren = true,
        visual = true,
        search = false,
      },
      transparent = false,
    },
  })
```
**Reproduce**

This issue happened when I update nightfox to a recent version.

## Additional info / context

nvim: 0.7 stable macOS arm64 installed via homebrew

**Comment by EdenEast** [comment]
I am unable to reproduce your issue. Is there a `minimal_init.lua` that will reproduce your issue? This is the current one I have to reproduce.

<details>
<summary>minimal_init.lua</summary>

```lua
-- `minimal_init.lua` used for reproducible configuration
-- Open with `nvim --clean -u minimal_init.lua`

local is_windows = vim.fn.has("win32") == 1
local function join(...)
  local sep = is_windows and "\\" or "/"
  return table.concat({ ... }, sep)
end

local root_tmp = is_windows and os.getenv("TEMP") or "/tmp"
local site_path = join(root_tmp, "nvim", "site")
local pack_path = join(site_path, "pack")
local install_path = join(pack_path, "packer", "start", "packer.nvim")
local compile_path = join(install_path, "plugin", "packer_compiled.lua")
vim.opt.packpath = site_path

local function load_plugins()
  local packer = require("packer")
  local use = packer.use
  packer.reset()
  packer.init({ compile_path = compile_path, package_root = pack_path })

  use("wbthomason/packer.nvim")
  use("edeneast/nightfox.nvim")
  use("nvim-treesitter/nvim-treesitter")
  -- ADD PLUGINS THAT ARE _NECESSARY_ FOR REPRODUCING THE ISSUE

  packer.install()
end

_G.load_config = function()
  -- ADD INIT.LUA SETTINGS THAT ARE _NECESSARY_ FOR REPRODUCING THE ISSUE

  require("nvim-treesitter.configs").setup({
    ensure_installed = { "lua" },
    highlight = {
      enable = true,
    },
  })

  require("nightfox").setup({
    options = {
      dim_inactive = false,
      styles = {
        comments = "italic",
        functions = "italic",
        keywords = "bold,italic",
      },
      inverse = {
        match_paren = true,
        visual = true,
        search = false,
      },
      transparent = false,
    },
  })

  vim.cmd("colorscheme nightfox")
end

if vim.fn.isdirectory(install_path) == 0 then
  vim.fn.system({ "git", "clone", "--depth=1", "https://github.com/wbthomason/packer.nvim", install_path })
end
load_plugins()
vim.cmd([[autocmd User PackerComplete ++once echo "Ready!" | lua load_config()]])
```

</details>

**Comment by LC0117** [comment]
Thanks for the reply!
I tried the minimal_config and find that this bug only exists when I have another colorscheme(`colorscheme <other_theme>` in config) and then switch back to nightfox(`colorscheme nightfox` in cmdline).(`NightfoxClean` and `NightfoxCompile` are made sure before that)
But it works fine when explicitly call `colorscheme nightfox` in config. By doing so switching themes won't cause the bug.

**Comment by EdenEast** [comment]
I tried to add another colorscheme like [catppuccin](https://github.com/catppuccin/nvim) and set that as the colorscheme in the minimal_init config. Then switch to `nightfox` but still cannot reproduce.

**Comment by LC0117** [comment]
Well, I think this is relative to my config. It's hard to reproduce as I use lua function to config a plugin like this.

```lua
-- in a file
use({"theme_A", config = conf.theme_A})
use({"theme_B", config = conf.theme_B})

-- in another file that stores configs
local M = {}
function M.theme_A() ... end
function M.theme_B() ... end
return M
```
I installed `rose-pine` theme and `nightfox` theme. And I call `vim.cmd('colorscheme rose-pine')` in `rose-pine`'s config function.

**Comment by EdenEast** [comment]
This might be interesting to you. The way I manage my themes is with a cache file (`$HOME/.cache/nvim/theme.txt`). On startup I try and load this file if it exists and set the colorscheme if it exists. The relavant section of my config [theme.lua]. Since you are also using and settings your configuration in the `config` section of packer you need to make sure that the `packer_compiled` file is evaluated before trying to set your colorscheme. You can see how I do this in my [main.lua]. I have an interesting way of structuring my config you might find it similar to what you are doing.

[theme.lua]: https://github.com/EdenEast/nyx/blob/bbadfea/config/.config/nvim/lua/eden/core/theme.lua
[main.lua]: https://github.com/EdenEast/nyx/blob/bbadfea/config/.config/nvim/lua/eden/main.lua

**Comment by LC0117** [comment]
Thanks for the idea!

**Comment by gegoune** [comment]
@EdenEast Apologies for hijacking closed issue but didn't feel like such simple question justifies new one.

If I run compile command or call api, once cache file gets created, is it automatically loaded next time nvim starts? If so, am I suppose to not call `setup()` function again? I am trying to figure out some oddity/get autocompile on bufwrite event of file with `setup()` working.

**Comment by EdenEast** [comment]
The `compile` functionality is a way to speed up startup time so that you don't have to keep re-computing all of nightfox's inner tables. The compile file is   the result of nightfox's current settings defined into all the set highlight calls. Calling the `setup` function only adds to nightfox's internal overrides table. Only when applied with the `colorscheme` command will nightfox resolve the overrides and call all of the set highlights.

In the [main](https://github.com/EdenEast/nightfox.nvim/blob/803d4ae/lua/nightfox/main.lua#L52-L58) function that applies the highlight groups it first checks if the precompiled file exists. If it does exist it will execute that file, instead of iterating though all the palettes, specs, and groups + overrides. This speeds up nightfox significantly.

When you are making changes to your configuration you should remove this compiled file with `NightfoxClean` so that nightfox will apply the palette, specs, groups and all the overrides.

TLDR: You should still define your `setup` function in your config. Execute `NightfoxCompile` once you are happy with your changes to get a speed increase. If you want to apply new changes make sure the compiled file is cleared so that the setup changes will be applied.

Hope this helped to understand how the compiled file works.

**Comment by gegoune** [comment]
Thank you very much, it makes sense and describe what one would expect from such caching system.

Please convert to new issue if you think it's warranted.

I got:

```lua

nightfox.setup {
  groups = {
    all = {
      TelescopeBorder = { fg = 'bg0', bg = 'bg0' },
    },
  },
}

vim.api.nvim_cmd({
  cmd = 'colorscheme',
  args = { 'nightfox' },
}, {})
```

After saving file containing that setup call (which is `require()`d from my `init.lua`), exiting nvim, opening it again, running: clean and compile commands I get:
```shell
❯ grep TelescopeBorder nightfox/nightfox_compiled.lua
vim.api.nvim_set_hl(0, "TelescopeBorder", { link = "" })
```

Am I doing something wrong?

**Comment by EdenEast** [comment]
I am not able to reproduce your issue. Is there a minimal_init with reproduction steps. Are you on the latest?

**Comment by gegoune** [comment]
Thanks for looking into it. Will try to come up with reproducible steps and open new issue if I manage.
